File: src/main/java/com/coding/programmers/level1/키패드누르기.java
package com.coding.programmers.level1;


public class 키패드누르기 {
    public static void main(String[] args) {
        Solution sol = new Solution();
        String result = sol.solution(new int[]{1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5}, "right");
        System.out.println(result);
    }
    private static class Solution {
        private static final int[] KEYPAD = {4,1,1,1,2,2,2,3,3,3};
        public String solution(int[] numbers, String hand) {
            StringBuilder answer = new StringBuilder();
            int right = -1;
            int left = -1;

            for (int number : numbers) {
                if (number == 1 || number == 4 || number == 7) {
                    left = number;
                    answer.append("L");
                } else if (number == 3 || number == 6 || number == 9) {
                    right = number;
                    answer.append("R");
                } else {
                    int l = keyLen(left, number);
                    int r = keyLen(right, number);
                    System.out.println("Left=" + left +", right=" + right + ", num="+ number + ", KL=" + l + ", KR=" + r);

                    if (l == r) {
                        if (hand.equals("right")) {
                            right = number;
                            answer.append("R");
                        } else {
                            left = number;
                            answer.append("L");
                        }
                    } else {
                        if (l > r) {
                            right = number;
                            answer.append("R");
                        } else {
                            left = number;
                            answer.append("L");
                        }
                    }
                }
            }

            return answer.toString();
        }

        private int keyLen(int num, int point) {
            System.out.println("pint=" + point + ", num=" + num);
            int sum = KEYPAD[num];
            if (num != 2 && num != 5 && num != 8 && num != 0) {
                sum += 1;
            }
            int result = Math.abs(KEYPAD[point] - sum);
            return result == 0 ? KEYPAD[point] : result;
        }
    }
}

File: src/main/kotlin/com/coding/programmers/implementation/키패드누르기.kt
package com.coding.programmers.implementation

import kotlin.math.abs

fun main() {
    val result = 키패드누르기().solution(
        intArrayOf(1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5), "right"
    )
    println(result)
}

/**
 * LRLLLRLLRRL
 * LRLLRRLLLRR
 */
class 키패드누르기 {
    fun solution(numbers: IntArray, hand: String): String {
        var answer = ""
        var left = 10
        var right = 12

        for (number in numbers) {
            if (number == 1 || number == 4 || number == 7) {
                left = number
                answer+= "L"
            } else if (number == 3 || number == 6 || number == 9) {
                right = number
                answer+= "R"
            } else {
                val leftHand = find(number, left)
                val rightHand = find(number, right)
                println("$leftHand, $rightHand")
                if (leftHand == rightHand) {
                    if (hand == "left") {
                        left = number
                        answer+= "L"
                    } else {
                        right = number
                        answer+= "R"
                    }
                } else {
                    if (leftHand < rightHand) {
                        left = number
                        answer+= "L"
                    } else {
                        right = number
                        answer+= "R"
                    }
                }
            }
        }
        return answer
    }
    fun find(src: Int, hand: Int): Int {
        val num = if (src == 0) 11 else src
        val handNum = if (hand == 0) 11 else hand
        val checkNum = abs(num - handNum)
        println("$src, $hand, = $checkNum")
        return (checkNum / 3) + (checkNum % 3)
    }
}